### ISSUE TITLE
About the log file

### ISSUE THREAD

**Issue by jackie840129** [ifzhang/FairMOT#75]
Hi,
I wonder that would you be willing to share your training log file?
I can successfully start to train with your code, but I'm not sure whether the loss value represents it is trained successfully.

**Comment by ifzhang** [comment]
The log file of training dla34 model with all the training set look like this:
2020-02-15-00-43: epoch: 1 |loss 15.513214 | hm_loss 0.971040 | wh_loss 1.627151 | off_loss 0.218932 | id_loss 8.861677 | time 65.350000 | 
2020-02-15-01-48: epoch: 2 |loss 13.010505 | hm_loss 0.713852 | wh_loss 1.335269 | off_loss 0.213037 | id_loss 7.760556 | time 65.083333 | 
2020-02-15-02-53: epoch: 3 |loss 11.865484 | hm_loss 0.646402 | wh_loss 1.256667 | off_loss 0.209433 | id_loss 7.134808 | time 65.050000 | 
2020-02-15-03-58: epoch: 4 |loss 11.110803 | hm_loss 0.600976 | wh_loss 1.216772 | off_loss 0.206983 | id_loss 6.722054 | time 64.983333 | 
2020-02-15-05-03: epoch: 5 |loss 10.579780 | hm_loss 0.567116 | wh_loss 1.185166 | off_loss 0.205041 | id_loss 6.437119 | time 65.000000 | 
2020-02-15-06-08: epoch: 6 |loss 10.166180 | hm_loss 0.539309 | wh_loss 1.159968 | off_loss 0.203650 | id_loss 6.218240 | time 64.950000 | 
2020-02-15-07-13: epoch: 7 |loss 9.842612 | hm_loss 0.516220 | wh_loss 1.137964 | off_loss 0.201564 | id_loss 6.052705 | time 64.916667 | 
2020-02-15-08-18: epoch: 8 |loss 9.600882 | hm_loss 0.498779 | wh_loss 1.122825 | off_loss 0.200493 | id_loss 5.928093 | time 64.950000 | 
2020-02-15-09-23: epoch: 9 |loss 9.371149 | hm_loss 0.484478 | wh_loss 1.106868 | off_loss 0.199528 | id_loss 5.804837 | time 64.950000 | 
2020-02-15-10-28: epoch: 10 |loss 9.174512 | hm_loss 0.468256 | wh_loss 1.095889 | off_loss 0.197702 | id_loss 5.709824 | time 65.066667 | 
2020-02-15-11-33: epoch: 11 |loss 9.002740 | hm_loss 0.457179 | wh_loss 1.080726 | off_loss 0.196913 | id_loss 5.619387 | time 65.116667 | 
2020-02-15-12-38: epoch: 12 |loss 8.855867 | hm_loss 0.446090 | wh_loss 1.068043 | off_loss 0.196213 | id_loss 5.545653 | time 65.233333 | 
2020-02-15-13-43: epoch: 13 |loss 8.743602 | hm_loss 0.439045 | wh_loss 1.062684 | off_loss 0.195315 | id_loss 5.485955 | time 65.200000 | 
2020-02-15-14-49: epoch: 14 |loss 8.607773 | hm_loss 0.427955 | wh_loss 1.059337 | off_loss 0.194575 | id_loss 5.417965 | time 65.250000 | 
2020-02-15-15-54: epoch: 15 |loss 8.497688 | hm_loss 0.421275 | wh_loss 1.040975 | off_loss 0.193923 | id_loss 5.361321 | time 65.266667 | 
2020-02-15-16-59: epoch: 16 |loss 8.418802 | hm_loss 0.413551 | wh_loss 1.035061 | off_loss 0.193347 | id_loss 5.325899 | time 65.366667 | 
2020-02-15-18-05: epoch: 17 |loss 8.309748 | hm_loss 0.407477 | wh_loss 1.021225 | off_loss 0.192540 | id_loss 5.267969 | time 65.316667 | 
2020-02-15-19-12: epoch: 18 |loss 8.242471 | hm_loss 0.400570 | wh_loss 1.025253 | off_loss 0.192096 | id_loss 5.236346 | time 67.216667 | 
2020-02-15-20-26: epoch: 19 |loss 8.149896 | hm_loss 0.393637 | wh_loss 1.009684 | off_loss 0.191500 | id_loss 5.191776 | time 74.233333 | 
2020-02-15-21-31: epoch: 20 |loss 8.067919 | hm_loss 0.387910 | wh_loss 1.007419 | off_loss 0.190404 | id_loss 5.150091 | time 65.233333 | 
2020-02-15-22-37: epoch: 21 |loss 8.014020 | hm_loss 0.382318 | wh_loss 0.993919 | off_loss 0.190030 | id_loss 5.128652 | time 65.283333 | 
2020-02-15-23-42: epoch: 22 |loss 7.946899 | hm_loss 0.376485 | wh_loss 0.995388 | off_loss 0.189376 | id_loss 5.095784 | time 65.300000 | 
2020-02-16-00-47: epoch: 23 |loss 7.882443 | hm_loss 0.372878 | wh_loss 0.986445 | off_loss 0.189029 | id_loss 5.061463 | time 65.300000 | 
2020-02-16-01-53: epoch: 24 |loss 7.818915 | hm_loss 0.367658 | wh_loss 0.984230 | off_loss 0.188475 | id_loss 5.030346 | time 65.333333 | 
2020-02-16-02-58: epoch: 25 |loss 7.784748 | hm_loss 0.363461 | wh_loss 0.976539 | off_loss 0.188343 | id_loss 5.017779 | time 65.333333 | 
2020-02-16-04-03: epoch: 26 |loss 7.709940 | hm_loss 0.359424 | wh_loss 0.975819 | off_loss 0.187913 | id_loss 4.975524 | time 65.333333 | 
2020-02-16-05-09: epoch: 27 |loss 7.671132 | hm_loss 0.355881 | wh_loss 0.970735 | off_loss 0.186999 | id_loss 4.959413 | time 65.350000 | 
2020-02-16-06-14: epoch: 28 |loss 7.617671 | hm_loss 0.352309 | wh_loss 0.964245 | off_loss 0.186876 | id_loss 4.931665 | time 65.366667 | 
2020-02-16-07-20: epoch: 29 |loss 7.568459 | hm_loss 0.349983 | wh_loss 0.962071 | off_loss 0.186159 | id_loss 4.904478 | time 65.400000 | 
2020-02-16-08-25: epoch: 30 |loss 7.535845 | hm_loss 0.346663 | wh_loss 0.956220 | off_loss 0.186243 | id_loss 4.890158 | time 65.383333 |

**Comment by jackie840129** [comment]
Thanks!
So it is normal that the id_loss is above 4... (Because some normal classification task the CE loss will go down below 1)

**Comment by ifzhang** [comment]
In fact, I do not train the parameter of the 'id_classifier', so the CE loss is high. I find that using this way can get higher IDF1 score. You can also add the parameters of the 'id_classifier' to the optimizer and the CE loss can drop to < 1. Training the 'id_classifier' can get lower id switches but still lower IDF1 on the validation set, which is a little bit strange. I'm not sure whether training the parameters of 'id_classifier' can get better results on the test set of MOT challenge.

**Comment by jackie840129** [comment]
wow.. OK I got it. It sounds strange lol...
So it just like training the embedding head to fit the randomly initialized classifier...

**Comment by jackie840129** [comment]
Sorry, I have another question~
I know the "emb_scale" is used in the previous work "Towards Real-time MOT", but do you know why it needs this scaling factor? (I've asked them but they didn't reply lol)

**Comment by shupinghu** [comment]
> In fact, I do not train the parameter of the 'id_classifier', so the CE loss is high. I find that using this way can get higher IDF1 score. You can also add the parameters of the 'id_classifier' to the optimizer and the CE loss can drop to < 1. Training the 'id_classifier' can get lower id switches but still lower IDF1 on the validation set, which is a little bit strange. I'm not sure whether training the parameters of 'id_classifier' can get better results on the test set of MOT challenge.
-----------------------------------------------------------------------------------------------
Hi, Since I want to train the parameter of the 'id_classifier', how can I add the parameters of the 'id_classifier' to the optimizer?

**Comment by ifzhang** [comment]
You can uncomment this line:
https://github.com/ifzhang/FairMOT/blob/1851158a1bc025da7e6cb839ddef9d14e33b404a/src/lib/trains/base_trainer.py#L30

**Comment by shupinghu** [comment]
> You can uncomment this line:
> https://github.com/ifzhang/FairMOT/blob/1851158a1bc025da7e6cb839ddef9d14e33b404a/src/lib/trains/base_trainer.py#L30

------------------------------------------------------------------------------------------------------------------
Hi,
in mot.py:  self.classifier = nn.Linear(self.emb_dim, self.nID)
and nID is depends the train dataset, in the default train dataset, nID=14455
Does this mean that the tracker of the trained model can only support the track of 14455 people in the meanwhile?

**Comment by jackie840129** [comment]
> > You can uncomment this line:
> > https://github.com/ifzhang/FairMOT/blob/1851158a1bc025da7e6cb839ddef9d14e33b404a/src/lib/trains/base_trainer.py#L30
> 
> Hi,
> in mot.py: self.classifier = nn.Linear(self.emb_dim, self.nID)
> and nID is depends the train dataset, in the default train dataset, nID=14455
> Does this mean that the tracker using the trained model can only support the track of 14455 people in the meanwhile?

Hi @shupinghu 
(I'm not the author) When testing, the model only uses the embedding feature for matching people in tracking. 
Just like human Re-ID, the number of IDs in training and testing sets are different.

**Comment by shupinghu** [comment]
> > > You can uncomment this line:
> > > https://github.com/ifzhang/FairMOT/blob/1851158a1bc025da7e6cb839ddef9d14e33b404a/src/lib/trains/base_trainer.py#L30
> > 
> > 
> > Hi,
> > in mot.py: self.classifier = nn.Linear(self.emb_dim, self.nID)
> > and nID is depends the train dataset, in the default train dataset, nID=14455
> > Does this mean that the tracker using the trained model can only support the track of 14455 people in the meanwhile?
> 
> Hi @shupinghu
> (I'm not the author) When testing, the model only uses the embedding feature for matching people in tracking.
> Just like human Re-ID, the number of IDs in training and testing sets are different.

But in the train set, the ID loss is determined using the classifier in mot.py, so I think the IDs is determined using the classifier.
in the test set, we do not use this classifier to get the ID number? I think it is a little bit strange

**Comment by shupinghu** [comment]
> > > You can uncomment this line:
> > > https://github.com/ifzhang/FairMOT/blob/1851158a1bc025da7e6cb839ddef9d14e33b404a/src/lib/trains/base_trainer.py#L30
> > 
> > 
> > Hi,
> > in mot.py: self.classifier = nn.Linear(self.emb_dim, self.nID)
> > and nID is depends the train dataset, in the default train dataset, nID=14455
> > Does this mean that the tracker using the trained model can only support the track of 14455 people in the meanwhile?
> 
> Hi @shupinghu
> (I'm not the author) When testing, the model only uses the embedding feature for matching people in tracking.
> Just like human Re-ID, the number of IDs in training and testing sets are different.

Sorry....I'm raw hand in re-id field.......

**Comment by TimeChi** [comment]
> Sorry, I have another question~
> I know the "emb_scale" is used in the previous work "Towards Real-time MOT", but do you know why it needs this scaling factor? (I've asked them but they didn't reply lol)

I have the same question, do you know the answer now @jackie840129
